README
NAME
    Mojo::mysql - Mojolicious and Async MySQL

SYNOPSIS
      use Mojo::mysql;

      # Connect to a local database
      my $mysql = Mojo::mysql->strict_mode('mysql://username@/test');

      # Connect to a remote database
      my $mysql = Mojo::mysql->strict_mode('mysql://username:password@hostname/test');
      # MySQL >= 8.0:
      my $mysql = Mojo::mysql->strict_mode('mysql://username:password@hostname/test;mysql_ssl=1');

      # Create a table
      $mysql->db->query(
        'create table names (id integer auto_increment primary key, name text)');

      # Insert a few rows
      my $db = $mysql->db;
      $db->query('insert into names (name) values (?)', 'Sara');
      $db->query('insert into names (name) values (?)', 'Stefan');

      # Insert more rows in a transaction
      eval {
        my $tx = $db->begin;
        $db->query('insert into names (name) values (?)', 'Baerbel');
        $db->query('insert into names (name) values (?)', 'Wolfgang');
        $tx->commit;
      };
      say $@ if $@;

      # Insert another row and return the generated id
      say $db->query('insert into names (name) values (?)', 'Daniel')
        ->last_insert_id;

      # Use SQL::Abstract to generate queries for you
      $db->insert('names', {name => 'Isabel'});
      say $db->select('names', undef, {name => 'Isabel'})->hash->{id};
      $db->update('names', {name => 'Bel'}, {name => 'Isabel'});
      $db->delete('names', {name => 'Bel'});

      # Select one row at a time
      my $results = $db->query('select * from names');
      while (my $next = $results->hash) {
        say $next->{name};
      }

      # Select all rows blocking
      $db->query('select * from names')
        ->hashes->map(sub { $_->{name} })->join("\n")->say;

      # Select all rows non-blocking
      Mojo::IOLoop->delay(
        sub {
          my $delay = shift;
          $db->query('select * from names' => $delay->begin);
        },
        sub {
          my ($delay, $err, $results) = @_;
          $results->hashes->map(sub { $_->{name} })->join("\n")->say;
        }
      )->wait;

      # Concurrent non-blocking queries (synchronized with promises)
      my $now   = $db->query_p('select now() as now');
      my $names = $db->query_p('select * from names');
      Mojo::Promise->all($now, $names)->then(sub {
        my ($now, $names) = @_;
        say $now->[0]->hash->{now};
        say $_->{name} for $names->[0]->hashes->each;
      })->catch(sub {
        my $err = shift;
        warn "Something went wrong: $err";
      })->wait;

      # Send and receive notifications non-blocking
      $mysql->pubsub->listen(foo => sub {
        my ($pubsub, $payload) = @_;
        say "foo: $payload";
        $pubsub->notify(bar => $payload);
      });
      $mysql->pubsub->listen(bar => sub {
        my ($pubsub, $payload) = @_;
        say "bar: $payload";
      });
      $mysql->pubsub->notify(foo => 'MySQL rocks!');

      Mojo::IOLoop->start unless Mojo::IOLoop->is_running;

DESCRIPTION
    Mojo::mysql is a tiny wrapper around DBD::mysql that makes MySQL
    <http://www.mysql.org> a lot of fun to use with the Mojolicious
    <http://mojolicio.us> real-time web framework.

    Database and handles are cached automatically, so they can be reused
    transparently to increase performance. And you can handle connection
    timeouts gracefully by holding on to them only for short amounts of
    time.

      use Mojolicious::Lite;
      use Mojo::mysql;

      helper mysql =>
        sub { state $mysql = Mojo::mysql->strict_mode('mysql://sri:s3cret@localhost/db') };

      get '/' => sub {
        my $c  = shift;
        my $db = $c->mysql->db;
        $c->render(json => $db->query('select now() as time')->hash);
      };

      app->start;

    While all I/O operations are performed blocking, you can wait for long
    running queries asynchronously, allowing the Mojo::IOLoop event loop to
    perform other tasks in the meantime. Since database connections usually
    have a very low latency, this often results in very good performance.

    Every database connection can only handle one active query at a time,
    this includes asynchronous ones. So if you start more than one, they
    will be put on a waiting list and performed sequentially. To perform
    multiple queries concurrently, you have to use multiple connections.

      # Performed sequentially (10 seconds)
      my $db = $mysql->db;
      $db->query('select sleep(5)' => sub {...});
      $db->query('select sleep(5)' => sub {...});

      # Performed concurrently (5 seconds)
      $mysql->db->query('select sleep(5)' => sub {...});
      $mysql->db->query('select sleep(5)' => sub {...});

    All cached database handles will be reset automatically if a new process
    has been forked, this allows multiple processes to share the same
    Mojo::mysql object safely.

EVENTS
    Mojo::mysql inherits all events from Mojo::EventEmitter and can emit the
    following new ones.

  connection
      $mysql->on(connection => sub {
        my ($mysql, $dbh) = @_;
        ...
      });

    Emitted when a new database connection has been established.

ATTRIBUTES
    Mojo::mysql implements the following attributes.

  abstract
      $abstract = $mysql->abstract;
      $mysql    = $mysql->abstract(SQL::Abstract->new);

    SQL::Abstract object used to generate CRUD queries for
    Mojo::mysql::Database.

      # Generate statements and bind values
      my ($stmt, @bind) = $mysql->abstract->select('names');

  auto_migrate
      my $bool = $mysql->auto_migrate;
      $mysql   = $mysql->auto_migrate($bool);

    Automatically migrate to the latest database schema with "migrations",
    as soon as the first database connection has been established.

    Defaults to false.

  database_class
      $class = $mysql->database_class;
      $mysql = $mysql->database_class("MyApp::Database");

    Class to be used by "db", defaults to Mojo::mysql::Database. Note that
    this class needs to have already been loaded before "db" is called.

  dsn
      my $dsn = $mysql->dsn;
      $mysql  = $mysql->dsn('dbi:mysql:dbname=foo');

    Data Source Name, defaults to "dbi:mysql:dbname=test".

  max_connections
      my $max = $mysql->max_connections;
      $mysql  = $mysql->max_connections(3);

    Maximum number of idle database handles to cache for future use,
    defaults to 5.

  migrations
      my $migrations = $mysql->migrations;
      $mysql         = $mysql->migrations(Mojo::mysql::Migrations->new);

    Mojo::mysql::Migrations object you can use to change your database
    schema more easily.

      # Load migrations from file and migrate to latest version
      $mysql->migrations->from_file('/Users/sri/migrations.sql')->migrate;

    MySQL does not support nested transactions and DDL transactions. DDL
    statements cause implicit "COMMIT". "ROLLBACK" will be called if any
    step of migration script fails, but only DML statements after the last
    implicit or explicit "COMMIT" can be reverted. Not all MySQL storage
    engines (like "MYISAM") support transactions.

    This means database will most likely be left in unknown state if
    migration script fails. Use this feature with caution and remember to
    always backup your database.

  options
      my $options = $mysql->options;
      $mysql      = $mysql->options({mysql_use_result => 1});

    Options for database handles, defaults to activating
    "mysql_enable_utf8", "AutoCommit", "AutoInactiveDestroy" as well as
    "RaiseError" and deactivating "PrintError". Note that "AutoCommit" and
    "RaiseError" are considered mandatory, so deactivating them would be
    very dangerous.

    "mysql_auto_reconnect" is never enabled, Mojo::mysql takes care of dead
    connections.

    "AutoCommit" cannot not be disabled, use $db->begin to manage
    transactions.

    "RaiseError" is enabled for blocking and disabled in event loop for
    non-blocking queries.

    Note about "mysql_enable_utf8":

      The mysql_enable_utf8 sets the utf8 charset which only supports up to 3-byte
      UTF-8 encodings. mysql_enable_utf8mb4 (as of DBD::mysql 4.032) properly
      supports encoding unicode characters to up to 4 bytes, such as 𠜎. It means the
      connection charset will be utf8mb4 (supported back to at least mysql 5.5) and
      these unicode characters will be supported, but no other changes.

    See also <https://github.com/jhthorsen/mojo-mysql/pull/32>

  password
      my $password = $mysql->password;
      $mysql       = $mysql->password('s3cret');

    Database password, defaults to an empty string.

  pubsub
      my $pubsub = $mysql->pubsub;
      $mysql     = $mysql->pubsub(Mojo::mysql::PubSub->new);

    Mojo::mysql::PubSub object you can use to send and receive notifications
    very efficiently, by sharing a single database connection with many
    consumers.

      # Subscribe to a channel
      $mysql->pubsub->listen(news => sub {
        my ($pubsub, $payload) = @_;
        say "Received: $payload";
      });

      # Notify a channel
      $mysql->pubsub->notify(news => 'MySQL rocks!');

    Note that Mojo::mysql::PubSub should be considered an experiment!

  username
      my $username = $mysql->username;
      $mysql       = $mysql->username('batman');

    Database username, defaults to an empty string.

METHODS
    Mojo::mysql inherits all methods from Mojo::EventEmitter and implements
    the following new ones.

  db
      my $db = $mysql->db;

    Get Mojo::mysql::Database object for a cached or newly created database
    handle. The database handle will be automatically cached again when that
    object is destroyed, so you can handle connection timeouts gracefully by
    holding on to it only for short amounts of time.

  from_string
      $mysql = $mysql->from_string('mysql://user@/test');

    Parse configuration from connection string.

      # Just a database
      $mysql->from_string('mysql:///db1');

      # Username and database
      $mysql->from_string('mysql://batman@/db2');

      # Username, password, host and database
      $mysql->from_string('mysql://batman:s3cret@localhost/db3');

      # Username, domain socket and database
      $mysql->from_string('mysql://batman@%2ftmp%2fmysql.sock/db4');

      # Username, database and additional options
      $mysql->from_string('mysql://batman@/db5?PrintError=1&RaiseError=0');

  new
      my $mysql = Mojo::mysql->new;
      my $mysql = Mojo::mysql->new(%attrs);
      my $mysql = Mojo::mysql->new(\%attrs);

Changes
Revision history for perl distribution Mojo-mysql

1.27 2023-10-26T20:59:16
 - Fix not cleaning up during GLOBAL_PHASE is DESTRUCT
   Contributor: Lasse Løvik

1.26 2022-12-09T10:41:01
 - Add support for "MOJO_MYSQL_PREFER_DRIVER" in Makefile.PL
 - Add support for autodetecting DBD::MariaDB
 - Fix spelling mistakes #88
   Contributor: Lucas Tiago de Moraes
 - Specified Perl version
 - Updated basic repository files
 - Updated contributors list

1.25 2021-11-22T18:13:22+0900
 - Changed DBD::mysql is not a dependency if DBD::MariaDB 1.21 is already installed #88

1.24 2021-10-06T07:57:45+0900
 - Fixed last_insert_id() and warnings_count() for MariaDB #86
   Contributor: Dan Book

1.23 2021-09-11T10:36:17+0200
 - Fix version number issues #84
 - Internal changes regarding MariaDB/mysql attributes

1.21 2021-04-28T12:15:39+0900
 - Fix uninitialized warning with expand(1) #83
 - Removed delay() from tests and documentation

1.20 2020-09-05T13:27:52+0900
 - Mojo::mysql::PubSub is less susceptible to deadlocks/timeouts #80
   Contributor: Larry Leszczynski

1.19 2020-05-01T06:58:48+0900
 - Fix documentation for Mojo::mysql::close_idle_connections()
 - Add documentation for SQL::Abstract::mysql::where() #78
 - Made SQL::Abstract::mysql more compatible with SQL::Abstract #77
   Contributor: Matt S Trout

1.18 2019-12-01T09:41:10+0100
 - Add missing code for SELECT AS

1.17 2019-08-01T09:44:57+0200
 - Fix leaking $sth when used in blocking mode, fixes #66

1.16 2019-06-25T06:33:55+0200
 - Add DBI to prerequisites
   Contributor: Mohammad S Anwar

1.15 2019-04-22T06:41:34+0200
 - Add support for NATURAL JOIN and JOIN USING #59

1.14 2019-03-23T08:07:17+0100
 - Correct handling of fetchall and arrays in Results

1.13 2019-03-02T11:27:01+0800
 - Add support for DBD::MariaDB #47
 - Add missing code for SQL JOIN #56
   Contributor: Tekki
 - Made it clearer that PubSub is an experiment

1.12 2019-01-05T12:34:13+0900
 - Bumped Mojolicious version to 8.03
 - Bumped SQL::Abstract version to 1.86 #49

1.11 2018-12-18T19:27:08+0900
 - Avoid "Gathering async_query_in_flight results for the wrong handle" warning

1.10 2018-12-18T07:25:14+0900
 - Add SQL::Abstract::mysql
   Contributor: Rolf Stöckli

1.09 2018-11-27T09:32:01+0900
 - Fix MariaDB/MySQL incompatibility #41
   Contributor: Rolf Stöckli
 - Fix documentation mismatch regarding "mysql_client_found_rows" #42
   Contributor: Yuriy Zhilovets

1.08 2018-11-13T17:31:49+0900
 - Fix query() with callback returns $self
 - Fix Gathering async_query_in_flight results for the wrong handle bug
 - Add close_idle_connections to Mojo::mysql
 - Add support for working with JSON
 - Add tables method to Mojo::mysql::Database
 - Change database name is optional in constructor #38

1.07 2018-05-03T12:25:08+0200
 - Fix using "mysql_socket" instead of "host" when connecting to a unix socket #34
 - Allow constructor to take a single hashref #37

1.06 2018-02-27T19:32:40+0100
 - Changed from_string() to also accept Mojo::URL objects #36
   Contributor: Karl Rune Nilsen

1.05 2017-11-11T10:04:40+0800
 - Add delete_p(), insert_p(), query_p(), select_p() and update_p()

1.04 2017-08-14T19:22:33+0200
 - Documented "mysql_enable_utf8" v.s. "mysql_enable_utf8mb4" #32
 - Can pass on attributes to new()

1.03 2017-05-21T23:19:29+0200
 - Add ability to set types of query parameters #31
   Contributor: Dan Book

1.02 2017-05-15T20:34:01+0200
 - Fix utf8 handling in DBD::mysql 4.042
 - Prevent warnings when creating the mojo_migrations table #26
 - Add proper quoting of table and column names #30
 - Add warning when using Mojo::mysql::PubSub

1.01 2017-03-25T08:24:29+0100
 - Add strict_mode() method and constructor #29

1.00 2017-02-12T18:30:58+0100
 - Add support for generating queries with SQL::Abstract
 - Add abstract attribute to Mojo::Pg
 - Add delete, insert, select and update methods to Mojo::Pg::Database
 - Add database_class attribute to Mojo::mysql
 - Add results_class attribute to Mojo::mysql::Database
 - Improved contextual caller information in query error messages
 - Compatible with Mojolicious 7.18

0.14 2016-02-15T14:06:24+0100
 - Add Mojo::mysql::auto_migrate

0.13 2016-01-27T21:05:37+0100
 - Remove deprecrated do() method
 - Add finish() to Mojo::mysql::Results
 - Fix bug where non-blocking queries could get lost after the database
   connection got closed unexpectedly
   https://github.com/kraih/mojo-pg/commit/2165b8e1131f2a5044ec2aae1c0ba8a00232b7c8
 - Improved Mojo::mysql::Migrations to detect if the currently active version
   is greater than the latest version.
   https://github.com/kraih/mojo-pg/commit/92bc312e725042b748950b9c61319d0256d0004a

0.12 2015-05-02T17:55:13Z
 - Added module Mojo::mysql::PubSub.
 - Added pubsub attribute to Mojo::mysql.

0.11 2015-04-06T03:38:31Z
 - Fixed bug in Mojo::mysql::Migrations where migrations would sometimes be
   executed in the wrong order.

0.10 2015-04-05T23:32:03Z
 - Fixed bug in Mojo::mysql::Migrations where the latest version could not
   always be determined correctly. (Hernan Lopes)
 - Updated blog example application from Mojo::Pg

0.09 2015-03-29T18:29:35Z
 - Fixed Mojo::mysql::Migrations to allow delimiter in comments and quoted
   strings

lib/Mojo/mysql.pm
package Mojo::mysql;
use Mojo::Base 'Mojo::EventEmitter';

use Carp 'croak';
use DBI;
use File::Spec::Functions 'file_name_is_absolute';
use Mojo::mysql::Database;
use Mojo::mysql::Migrations;
use Mojo::URL;
use Scalar::Util 'weaken';
use SQL::Abstract::mysql;

our $VERSION = '1.27';

has abstract        => sub { SQL::Abstract::mysql->new(quote_char => chr(96), name_sep => '.') };
has auto_migrate    => 0;
has database_class  => 'Mojo::mysql::Database';
has dsn             => 'dbi:mysql:dbname=test';
has max_connections => 5;

has migrations => sub {
  my $migrations = Mojo::mysql::Migrations->new(mysql => shift);
  weaken $migrations->{mysql};
  return $migrations;
};

has options => sub {
  my $self    = shift;
  my $options = {AutoCommit => 1, AutoInactiveDestroy => 1, PrintError => 0, RaiseError => 1};
  $options->{mysql_enable_utf8} = 1 if $self->dsn =~ m!^dbi:mysql:!;
  return $options;
};

has [qw(password username)] => '';
has pubsub                  => sub {
  require Mojo::mysql::PubSub;
  my $pubsub = Mojo::mysql::PubSub->new(mysql => shift);
  warn "Use of Mojo::mysql::PubSub is highly EXPERIMENTAL and should be considered an experiment"
    unless $ENV{MOJO_PUBSUB_EXPERIMENTAL};
  weaken $pubsub->{mysql};
  return $pubsub;
};

sub close_idle_connections {
  my ($self, $keep) = (shift, $_[0] || 0);
  my $queue = $self->{queue} || [];

  # The database handle needs to be destroyed before the file handle
  shift(@$queue)->[0] = undef while @$queue > $keep;
  return $self;
}

sub db {
  my $self = shift;

  # Fork safety
  delete @$self{qw(pid queue)} unless ($self->{pid} //= $$) eq $$;

  my ($dbh, $handle) = @{$self->_dequeue};
  return $self->database_class->new(dbh => $dbh, handle => $handle, mysql => $self);
}

sub from_string {
  my ($self, $str) = @_;

  # Protocol
  return $self unless $str;
  my $url = UNIVERSAL::isa($str, 'Mojo::URL') ? $str : Mojo::URL->new($str);
  croak qq{Invalid MySQL/MariaDB connection string "$str"} unless $url->protocol =~ m!^(mariadb|mysql)$!;
  my $dsn = $url->protocol eq 'mariadb' ? 'dbi:MariaDB' : 'dbi:mysql';

  # https://github.com/jhthorsen/mojo-mysql/pull/47
  die "DBD::MariaDB 1.21 is required for Mojo::mysql to work properly"
    if $dsn eq 'dbi:MariaDB' and !eval 'use DBD::MariaDB 1.21;1';

  # Database
  $dsn .= ':dbname=' . $url->path->parts->[0] if defined $url->path->parts->[0];

  # Host and port
  if (my $host = $url->host) { $dsn .= file_name_is_absolute($host) ? ";mysql_socket=$host" : ";host=$host" }
  if (my $port = $url->port) { $dsn .= ";port=$port" }

  # Need to set the dsn before reading options
  $self->dsn($dsn);

  # Username and password
  if (($url->userinfo // '') =~ /^([^:]+)(?::([^:]+))?$/) {
    $self->username($1);
    $self->password($2) if defined $2;
  }

  # Options
  my $hash = $url->query->to_hash;
  @{$self->options}{keys %$hash} = values %$hash;

  return $self;
}

sub new {
  @_ > 2 || ref $_[-1] eq 'HASH' ? shift->SUPER::new(@_) : shift->SUPER::new->from_string(@_);
}

sub strict_mode {
  my $self = ref $_[0] ? shift : shift->new(@_);
  $self->{strict_mode} = $_[0] ? 1 : @_ ? 0 : 1;
  warn "[Mojo::mysql] strict_mode($self->{strict_mode})\n" if $ENV{DBI_TRACE};
  $self->close_idle_connections;
  return $self;
}

sub _dequeue {
  my $self = shift;
  my $dbh;

  while (my $c = shift @{$self->{queue}}) { return $c if $c->[0]->ping }
  $dbh = DBI->connect(map { $self->$_ } qw(dsn username password options));

  # <mst> batman's probably going to have more "fun" than you have ...
  # especially once he discovers that DBD::mysql randomly reconnects under
  # you, silently, but only if certain env vars are set
  # hint: force-set mysql_auto_reconnect or whatever it's called to 0
  Mojo::mysql::Database->_dbh_attr($dbh, mysql_auto_reconnect => 0);

  # Maintain Commits with Mojo::mysql::Transaction
  $dbh->{AutoCommit} = 1;

  $self->_set_strict_mode($dbh) if $self->{strict_mode};
  $self->migrations->migrate    if $self->auto_migrate and !$self->{migrated}++;
  $self->emit(connection => $dbh);
  [$dbh];
}

sub _enqueue {
  my ($self, $dbh, $handle) = @_;
  push @{$self->{queue}}, [$dbh, $handle] if $dbh->{Active};
  $self->close_idle_connections($self->max_connections);
}

sub _set_strict_mode {
  $_[1]->do(q[SET SQL_MODE = CONCAT('ANSI,TRADITIONAL,ONLY_FULL_GROUP_BY,', @@sql_mode)]);
  $_[1]->do(q[SET SQL_AUTO_IS_NULL = 0]);
}

1;

=encoding utf8

=head1 NAME

Mojo::mysql - Mojolicious and Async MySQL/MariaDB

=head1 SYNOPSIS

  use Mojo::mysql;

  # Connect to a local database
  my $mysql = Mojo::mysql->strict_mode('mysql://username@/test');

  # Connect to a remote database
  my $mysql = Mojo::mysql->strict_mode('mysql://username:password@hostname/test');
  # MySQL >= 8.0:
  my $mysql = Mojo::mysql->strict_mode('mysql://username:password@hostname/test;mysql_ssl=1');

  # Use DBD::MariaDB instead of DBD::mysql
  my $mysql = Mojo::mysql->strict_mode('mariadb://username@/test');

  # Create a table
  $mysql->db->query(
    'create table names (id integer auto_increment primary key, name text)');

  # Insert a few rows
  my $db = $mysql->db;
  $db->query('insert into names (name) values (?)', 'Sara');
  $db->query('insert into names (name) values (?)', 'Stefan');

  # Insert more rows in a transaction
  eval {
    my $tx = $db->begin;
    $db->query('insert into names (name) values (?)', 'Baerbel');
    $db->query('insert into names (name) values (?)', 'Wolfgang');
    $tx->commit;
  };
  say $@ if $@;

  # Insert another row and return the generated id
  say $db->query('insert into names (name) values (?)', 'Daniel')
    ->last_insert_id;

  # Use SQL::Abstract::mysql to generate queries for you
  $db->insert('names', {name => 'Isabel'});
  say $db->select('names', undef, {name => 'Isabel'})->hash->{id};
  $db->update('names', {name => 'Bel'}, {name => 'Isabel'});
  $db->delete('names', {name => 'Bel'});

  # Select one row at a time
  my $results = $db->query('select * from names');
  while (my $next = $results->hash) {
    say $next->{name};
  }

  # Select all rows blocking
  $db->query('select * from names')
    ->hashes->map(sub { $_->{name} })->join("\n")->say;

  # Select all rows non-blocking
  $db->query('select * from names' => sub {
    my ($db, $err, $results) = @_;
    $results->hashes->map(sub { $_->{name} })->join("\n")->say;
  });

  # Concurrent non-blocking queries (synchronized with promises)
  my $now   = $db->query_p('select now() as now');
  my $names = $db->query_p('select * from names');
  Mojo::Promise->all($now, $names)->then(sub {
    my ($now, $names) = @_;
    say $now->[0]->hash->{now};
    say $_->{name} for $names->[0]->hashes->each;
  })->catch(sub {
    my $err = shift;
    warn "Something went wrong: $err";
  })->wait;

  Mojo::IOLoop->start unless Mojo::IOLoop->is_running;

=head1 DESCRIPTION

L<Mojo::mysql> is a tiny wrapper around L<DBD::mysql> and L<DBD::MariaDB> that
makes L<MySQL|http://www.mysql.org> and L<MariaDB|https://mariadb.org/> a lot
of fun to use with the L<Mojolicious|http://mojolicio.us> real-time web
framework.

The two DBD drivers are compatible with both MySQL and MariaDB, but they offer
different L</options>. L<DBD::MariaDB> should have better unicode support
though and might become the default in the future.

Database and handles are cached automatically, so they can be reused
transparently to increase performance. And you can handle connection timeouts
gracefully by holding on to them only for short amounts of time.

  use Mojolicious::Lite;
  use Mojo::mysql;

  helper mysql =>
    sub { state $mysql = Mojo::mysql->strict_mode('mysql://sri:s3cret@localhost/db') };

  get '/' => sub {
    my $c  = shift;
    my $db = $c->mysql->db;
    $c->render(json => $db->query('select now() as time')->hash);
  };

  app->start;

While all I/O operations are performed blocking, you can wait for long running
queries asynchronously, allowing the L<Mojo::IOLoop> event loop to perform
other tasks in the meantime. Since database connections usually have a very low
latency, this often results in very good performance.

Every database connection can only handle one active query at a time, this
includes asynchronous ones. So if you start more than one, they will be put on
a waiting list and performed sequentially. To perform multiple queries
concurrently, you have to use multiple connections.

  # Performed sequentially (10 seconds)
  my $db = $mysql->db;
  $db->query('select sleep(5)' => sub {...});
  $db->query('select sleep(5)' => sub {...});

  # Performed concurrently (5 seconds)
  $mysql->db->query('select sleep(5)' => sub {...});
  $mysql->db->query('select sleep(5)' => sub {...});

All cached database handles will be reset automatically if a new process has
been forked, this allows multiple processes to share the same L<Mojo::mysql>
object safely.

=head1 EVENTS

L<Mojo::mysql> inherits all events from L<Mojo::EventEmitter> and can emit the
following new ones.

=head2 connection

  $mysql->on(connection => sub {
    my ($mysql, $dbh) = @_;
    ...
  });

Emitted when a new database connection has been established.

=head1 ATTRIBUTES

L<Mojo::mysql> implements the following attributes.

=head2 abstract

  $abstract = $mysql->abstract;
  $mysql    = $mysql->abstract(SQL::Abstract::mysql->new);

L<SQL::Abstract::mysql> object used to generate CRUD queries for L<Mojo::mysql::Database>.

  # Generate statements and bind values
  my ($stmt, @bind) = $mysql->abstract->select('names');

=head2 auto_migrate

  my $bool = $mysql->auto_migrate;
  $mysql   = $mysql->auto_migrate($bool);

Automatically migrate to the latest database schema with L</"migrations">, as
soon as the first database connection has been established.

Defaults to false.

=head2 database_class

  $class = $mysql->database_class;
  $mysql = $mysql->database_class("MyApp::Database");

Class to be used by L</"db">, defaults to L<Mojo::mysql::Database>. Note that this
class needs to have already been loaded before L</"db"> is called.

=head2 dsn

  my $dsn = $mysql->dsn;
  $mysql  = $mysql->dsn('dbi:mysql:dbname=foo');

Data Source Name, defaults to C<dbi:mysql:dbname=test>.

=head2 max_connections

  my $max = $mysql->max_connections;
  $mysql  = $mysql->max_connections(3);

Maximum number of idle database handles to cache for future use, defaults to
C<5>.

=head2 migrations

  my $migrations = $mysql->migrations;
  $mysql         = $mysql->migrations(Mojo::mysql::Migrations->new);

L<Mojo::mysql::Migrations> object you can use to change your database schema more
easily.

  # Load migrations from file and migrate to latest version
  $mysql->migrations->from_file('/Users/sri/migrations.sql')->migrate;

MySQL and MariaDB does not support nested transactions and DDL transactions.
DDL statements cause implicit C<COMMIT>. C<ROLLBACK> will be called if any step
of migration script fails, but only DML statements after the last implicit or
explicit C<COMMIT> can be reverted. Not all storage engines (like C<MYISAM>)
support transactions.

This means database will most likely be left in unknown state if migration script fails.
Use this feature with caution and remember to always backup your database.

=head2 options

  my $options = $mysql->options;
  $mysql      = $mysql->options({mysql_use_result => 1});

Options for database handles, defaults to activating C<mysql_enable_utf8> (only
for L<DBD::mysql>), C<AutoCommit>, C<AutoInactiveDestroy> as well as
C<RaiseError> and deactivating C<PrintError>. C<AutoCommit> and C<RaiseError>
are considered mandatory, so deactivating them would be very dangerous.

C<mysql_auto_reconnect> is never enabled, L<Mojo::mysql> takes care of dead connections.

C<AutoCommit> cannot not be disabled, use $db->L<begin|Mojo::mysql::Database/"begin"> to manage transactions.

C<RaiseError> is enabled for blocking and disabled in event loop for non-blocking queries.

About C<mysql_enable_utf8>:

  The mysql_enable_utf8 sets the utf8 charset which only supports up to 3-byte
  UTF-8 encodings. mysql_enable_utf8mb4 (as of DBD::mysql 4.032) properly
  supports encoding unicode characters to up to 4 bytes, such as 𠜎. It means the
  connection charset will be utf8mb4 (supported back to at least mysql 5.5) and
  these unicode characters will be supported, but no other changes.

See also L<https://github.com/jhthorsen/mojo-mysql/pull/32>

=head2 password

  my $password = $mysql->password;
  $mysql       = $mysql->password('s3cret');

Database password, defaults to an empty string.

=head2 pubsub

  my $pubsub = $mysql->pubsub;
  $mysql     = $mysql->pubsub(Mojo::mysql::PubSub->new);

L<Mojo::mysql::PubSub> should be considered an EXPERIMENT! See
L<Mojo::mysql::PubSub/DESCRIPTION> for more information.

=head2 username

  my $username = $mysql->username;
  $mysql       = $mysql->username('batman');

Database username, defaults to an empty string.

=head1 METHODS

L<Mojo::mysql> inherits all methods from L<Mojo::EventEmitter> and implements the
following new ones.

=head2 close_idle_connections

  $mysql = $mysql->close_idle_connections($keep);

Close all connections that are not currently active, or limit the
number of idle connections to C<$keep>.

=head2 db

  my $db = $mysql->db;

Get L<Mojo::mysql::Database> object for a cached or newly created database
handle. The database handle will be automatically cached again when that
object is destroyed, so you can handle connection timeouts gracefully by
holding on to it only for short amounts of time.

=head2 from_string

  $mysql = $mysql->from_string('mysql://user@/test');

Parse configuration from connection string.

  # Just a database
  $mysql->from_string('mysql:///db1');

  # Username and database
  $mysql->from_string('mysql://batman@/db2');

  # Username, password, host and database
  $mysql->from_string('mysql://batman:s3cret@localhost/db3');

  # Username, domain socket and database
  $mysql->from_string('mysql://batman@%2ftmp%2fmysql.sock/db4');

  # Username, database and additional options
  $mysql->from_string('mysql://batman@/db5?PrintError=1&RaiseError=0');

=head2 new

  my $mysql = Mojo::mysql->new;
  my $mysql = Mojo::mysql->new(%attrs);
  my $mysql = Mojo::mysql->new(\%attrs);
  my $mysql = Mojo::mysql->new('mysql://user@/test');
  my $mysql = Mojo::mysql->new('mariadb://user@/test');

Construct a new L<Mojo::mysql> object either from L</ATTRIBUTES> and or parse
connection string with L</"from_string"> if necessary.

Using the "mariadb" scheme requires the optional module L<DBD::MariaDB> version
1.21 (or later) to be installed.

=head2 strict_mode

  my $mysql = Mojo::mysql->strict_mode('mysql://user@/test');
  my $mysql = $mysql->strict_mode($boolean);

This method can act as both a constructor and a method. When called as a
constructor, it will be the same as:

  my $mysql = Mojo::mysql->new('mysql://user@/test')->strict_mode(1);

Enabling strict mode will execute the following statement when a new connection
is created:

  SET SQL_MODE = CONCAT('ANSI,TRADITIONAL,ONLY_FULL_GROUP_BY,', @@sql_mode)
  SET SQL_AUTO_IS_NULL = 0

The idea is to set up a connection that makes it harder for MySQL to allow
"invalid" data to be inserted.

This method will not be removed, but the internal commands is subject to
change.

=head1 DEBUGGING

You can set the C<DBI_TRACE> environment variable to get some advanced
diagnostics information printed to C<STDERR> by L<DBI>.

  DBI_TRACE=1
  DBI_TRACE=15
  DBI_TRACE=15=dbitrace.log
  DBI_TRACE=SQL
  DBI_PROFILE=2

See also L<https://metacpan.org/pod/DBI#DBI_TRACE> and
L<https://metacpan.org/pod/DBI#DBI_PROFILE>.

=head1 REFERENCE

This is the class hierarchy of the L<Mojo::mysql> distribution.

=over 2

=item * L<Mojo::mysql>

=item * L<Mojo::mysql::Database>

=item * L<Mojo::mysql::Migrations>

=item * L<Mojo::mysql::PubSub>

=item * L<Mojo::mysql::Results>

=item * L<Mojo::mysql::Transaction>

=back

=head1 AUTHORS

This project is highly inspired by Sebastian Riedel's L<Mojo::Pg>.

=head2 Project Founder

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=head2 Contributors

=over 2


=item * Adam Hopkins <violapiratejunky@gmail.com>

=item * Alexander Karelas <alex.karelas@gmail.com>

=item * Curt Hochwender <hochwender@centurytel.net>

=item * Dan Book <grinnz@grinnz.com>

=item * Doug Bell <doug@preaction.me>

=item * Florian Heyer <info@heyer-it.de>

=item * Hernan Lopes <hernanlopes@gmail.com>

=item * Karl Rune Nilsen <krunen@gmail.com>

=item * Larry Leszczynski <larryl@emailplus.org>

=item * Lucas Tiago de Moraes <lucastiagodemoraes@gmail.com>

=item * Matt S Trout <mst@shadowcat.co.uk>

=item * Mike Magowan <mike@magowan.co.uk>

=item * Mohammad S Anwar <mohammad.anwar@yahoo.com>

=item * Rolf Stöckli <tekki@cpan.org>

=item * Sebastian Riedel <sri@cpan.org>

=item * Svetoslav Naydenov <harry@bix.bg>

=item * Svetoslav Naydenov <harryl@cpan.org>

=item * Tekki <tekki@tekki.ch>

=back

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014-2019, Jan Henning Thorsen.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<https://github.com/jhthorsen/mojo-mysql>,

L<Mojo::Pg> Async Connector for PostgreSQL using L<DBD::Pg>, L<https://github.com/kraih/mojo-pg>,

L<Mojo::MySQL5> Pure-Perl non-blocking I/O MySQL Connector, L<https://github.com/harry-bix/mojo-mysql5>,

L<Mojolicious::Guides>, L<http://mojolicio.us>.

=cut
